gandalf.js

```diff
@@ -1,7 +1,7 @@
 /**
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?